src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    internal static class RouteParameterParser
```

Comment:
This is not a great place to start reviewing. This code is a virtual cut/paste from existing code - and some of it is total spaghetti (not my doing lol).

I relayered the existing parser stuff to use the new stuff.

Reply:
I don't really think so since we ultimately need to surface all of the substrings we create as `string`. Also this code runs at startup, if this is on the critical path for anyone then we haven't done our job 😆

Reply:
> if this is on the critical path for anyone then we haven't done our job